src/ott/neural/flows/genot.py

```diff
@@ -0,0 +1,490 @@
+            b=tmat.sum(axis=0),
+            state_eta=self.state_eta,
+            state_xi=self.state_xi,
+        )
+      if iteration % self.valid_freq == 0:
+        self._valid_step(valid_loader, iteration)
+        if self.checkpoint_manager is not None:
+          states_to_save = {"state_velocity_field": self.state_velocity_field}
+          if self.state_eta is not None:
+            states_to_save["state_eta"] = self.state_eta
+          if self.state_xi is not None:
+            states_to_save["state_xi"] = self.state_xi
+          self.checkpoint_manager.save(iteration, states_to_save)
+
+  def _get_step_fn(self) -> Callable:
+
+    @jax.jit
+    def step_fn(
+        key: jax.random.PRNGKeyArray,
+        state_velocity_field: train_state.TrainState,
+        batch: Dict[str, jnp.array],
+    ):
+
+      def loss_fn(
+          params: jnp.ndarray, batch: Dict[str, jnp.array],
+          rng: jax.random.PRNGKeyArray
+      ):
+        x_t = self.flow.compute_xt(
+            batch["noise"], batch["time"], batch["latent"], batch["target"]
+        )
+        apply_fn = functools.partial(
+            state_velocity_field.apply_fn, {"params": params}
+        )
+
+        cond_input = jnp.concatenate([
+            batch[el]
+            for el in ["source", "source_conditions"]
+            if batch[el] is not None
+        ],
+                                     axis=1)
+        v_t = jax.vmap(apply_fn
+                      )(t=batch["time"], x=x_t, condition=cond_input, rng=rng)
+        u_t = self.flow.compute_ut(
+            batch["time"], batch["latent"], batch["target"]
+        )
+        return jnp.mean((v_t - u_t) ** 2)
+
+      keys_model = jax.random.split(key, len(batch["noise"]))
+
+      grad_fn = jax.value_and_grad(loss_fn, has_aux=False)
+      loss, grads = grad_fn(state_velocity_field.params, batch, keys_model)
+
+      return state_velocity_field.apply_gradients(grads=grads), loss
+
+    return step_fn
+
+  def transport(
+      self,
+      source: jnp.ndarray,
+      condition: Optional[jnp.ndarray] = None,
+      rng: Optional[jax.Array] = None,
+      forward: bool = True,
+      diffeqsolve_kwargs: Dict[str, Any] = types.MappingProxyType({}),
+  ) -> Union[jnp.array, diffrax.Solution, Optional[jnp.ndarray]]:
+    """Transport data with the learnt plan.
+
+    This method pushes-forward the `source` to its conditional distribution by
+      solving the neural ODE parameterized by the
+      :attr:`~ott.neural.flows.genot.velocity_field`
+
+    Args:
+      source: Data to transport.
+      condition: Condition of the input data.
+      rng: random seed for sampling from the latent distribution.
+      forward: If `True` integrates forward, otherwise backwards.
+      diffeqsolve_kwargs: Keyword arguments for the ODE solver.
+
+    Returns:
+      The push-forward or pull-back distribution defined by the learnt
+      transport plan.
+
+    """
+    rng = utils.default_prng_key(rng)
+    if not forward:
```

Comment:
Are you planning implementing the backward case? If not, would remove the argument.

Reply:
Yeah not entirely sure whether it makes sense, because you'll end up in a normal distribution. But for having a uniform API , it might make sense to have it?